fontir/src/variations.rs

```diff
@@ -0,0 +1,632 @@
+
+    use crate::{
+        coords::{NormalizedCoord, NormalizedLocation},
+        variations::{Tent, ONE},
+    };
+
+    use super::{VariationModel, VariationRegion};
+
+    fn norm_loc(positions: &[(&str, f32)]) -> NormalizedLocation {
+        let mut loc = NormalizedLocation::new();
+        for (axis, value) in positions {
+            loc.set_pos(axis.to_owned(), NormalizedCoord::new(*value));
+        }
+        return loc;
+    }
+
+    fn tent(lower: f32, peak: f32, upper: f32) -> Tent {
+        Tent {
+            lower: NormalizedCoord::new(lower),
+            peak: NormalizedCoord::new(peak),
+            upper: NormalizedCoord::new(upper),
+        }
+    }
+
+    /// Python
+    /// >>> supportScalar({}, {})
+    /// 1.0
+    #[test]
+    fn multiplier_at_default_for_default() {
+        let loc = NormalizedLocation::new();
+        assert_eq!(Some(ONE), VariationRegion::new().multiplier_at(&loc));
+    }
+
+    /// Python
+    /// >>> supportScalar({'wght':.2}, {})
+    /// 1.0
+    #[test]
+    fn multiplier_at_off_default_for_default() {
+        let loc = norm_loc(&[("Weight", 0.2)]);
+        assert_eq!(Some(ONE), VariationRegion::new().multiplier_at(&loc));
+    }
+
+    /// Python
+    /// >>> supportScalar({'wght':.2}, {'wght':(0,2,3)})
+    /// 0.1
+    #[test]
+    fn multiplier_at_off_default_for_simple_weight() {
+        let loc = norm_loc(&[("Weight", 0.2)]);
+        let mut region = VariationRegion::new();
+        region.0.insert("Weight".to_string(), tent(0.0, 2.0, 3.0));
+        assert_eq!(Some(OrderedFloat(0.1)), region.multiplier_at(&loc));
+    }
+
+    /// Python
+    /// >>> supportScalar({'wght':2.5}, {'wght':(0,2,4)})
+    /// 0.75
+    #[test]
+    fn multiplier_at_for_weight() {
+        let loc = norm_loc(&[("Weight", 2.5)]);
+        let mut region = VariationRegion::new();
+        region.0.insert("Weight".to_string(), tent(0.0, 2.0, 4.0));
+        assert_eq!(Some(OrderedFloat(0.75)), region.multiplier_at(&loc));
+    }
+
+    /// Python
+    /// >>> supportScalar({'wght':2.5, 'wdth':0}, {'wght':(0,2,4), 'wdth':(-1,0,+1)})
+    /// 0.75
+    /// Note that under font rules a peak of 0 means no influence
+    #[test]
+    fn multiplier_at_for_weight_width() {
+        let loc = norm_loc(&[("Weight", 2.5), ("Width", 0.0)]);
+        let mut region = VariationRegion::new();
+        region.0.insert("Weight".to_string(), tent(0.0, 2.0, 4.0));
+        region.0.insert("Width".to_string(), tent(-1.0, 0.0, 1.0));
+        assert_eq!(Some(OrderedFloat(0.75)), region.multiplier_at(&loc));
+    }
+
+    /// Python has several examples that are really just checking the width peak 0 does nothing
+    /// >>> supportScalar({'wght':2.5, 'wdth':0}, {'wght':(0,2,4), 'wdth':(-1,0,+1)})
+    /// 0.75
+    /// >>> supportScalar({'wght':2.5, 'wdth':.5}, {'wght':(0,2,4), 'wdth':(-1,0,+1)})
+    /// 0.75
+    #[test]
+    fn zero_peak_means_no_impact() {
+        assert_eq!(false, tent(-1.0, 0.0, 1.0).validate());
```

Comment:
i'm not sure if "invalid" is the right word for this.. a peak at 0.0 simply means that the axis associated with the given tent does not participate in the delta interpolation in that particular region, i.e. per-axis support scalar is 1.0 so it has no effect when multiplied with other scalars or deltas.

Reply:
In FontTools non-OT tents are supported and tents like this can make sense. To my understanding in OT this form of tent is correctly described as invalid, it doesn't make sense and the code filters them out when they are encountered in OT mode. In Rust I only support OT mode so they are always invalid.

Or ... did I simply misunderstand the python?

Reply:
honestly I have never seen supportScalar being used anywhere with anything other than ot=True default...

> To my understanding in OT this form of tent is correctly described as invalid

hm I'd argue that a peak-0 axis tent is the _default_.. In fonttools code for gvar's TupleVariations, a tuple's region is represented by a dict keyed by axis tag, with values being (min,peak,max) tent triplets; when an axis doesn't participate within that region (i.e. peak == 0) then it is omitted from the dict; when compiling, the omitted axis tent is assumed to default to (0.0, 0.0, 0.0).

actually, one can potentially define a region where _all_ the axis tents peak at 0, and that basically means "please always apply my deltas at full scale (scalar=1.0) no matter what instance location"... And to my understanding that is not invalid. It's just the way things works in OTvar.

Maybe rename `Tent::validate` method to something like `does_participate` or something

Reply:
The argument for invalid is that it appears to me it will *never* participate, it will just *always* be filtered out. If it had influence on some regions and not others then does_participate makes more sense to me, but perhaps I misunderstand.

Reply:
See in particular the `if ot:` series of continue's at https://github.com/fonttools/fonttools/blob/2f1f5e5e7be331d960a0e30d537c2b4c70d89285/Lib/fontTools/varLib/models.py#L160. Tents that are so formed will *never* have influence. My argument is that this makes them illformed or invalid, whatever the user thought their sources would achieve probably isn't being achieved.

Is that fair or not so much?

Reply:
Hm. In debugging I think it may be that taking (0,0,0) to mean 1.0 might actually simplify things. Let me play with that more.